core/src/main/java/org/apache/iceberg/RollingManifestWriter.java

```diff
@@ -0,0 +1,243 @@
+    this.snapshotId = snapshotId;
+    this.spec = spec;
+    this.io = io;
+    this.outputFileFactory = outputFileFactory;
+    this.targetFileSizeInBytes = targetFileSizeInBytes;
+    this.manifestFiles = Lists.newArrayList();
+
+    openCurrentWriter();
+  }
+
+  protected abstract ManifestWriter<T> newManifestWriter(
+      int targetFormatVersion,
+      Long targetSnapshotId,
+      PartitionSpec targetSpec,
+      OutputFile targetManifestPath);
+
+  /**
+   * Add an added entry for a file.
+   *
+   * <p>The entry's snapshot ID will be this manifest's snapshot ID. The data and file sequence
+   * numbers will be assigned at commit.
+   *
+   * @param addedFile a data file
+   */
+  @Override
+  public void add(T addedFile) {
+    currentWriter.add(addedFile);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add an added entry for a file with a specific sequence number.
+   *
+   * <p>The entry's snapshot ID will be this manifest's snapshot ID. The entry's data sequence
+   * number will be the provided data sequence number. The entry's file sequence number will be
+   * assigned at commit.
+   *
+   * @param addedFile a data file
+   * @param dataSequenceNumber a data sequence number for the file
+   */
+  public void add(T addedFile, long dataSequenceNumber) {
+    currentWriter.add(addedFile, dataSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add an existing entry for a file.
+   *
+   * <p>The original data and file sequence numbers, snapshot ID, which were assigned at commit,
+   * must be preserved when adding an existing entry.
+   *
+   * @param existingFile a file
+   * @param fileSnapshotId snapshot ID when the data file was added to the table
+   * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @param fileSequenceNumber a file sequence number (assigned when the file was added)
+   */
+  public void existing(
+      T existingFile, long fileSnapshotId, long dataSequenceNumber, Long fileSequenceNumber) {
+    currentWriter.existing(existingFile, fileSnapshotId, dataSequenceNumber, fileSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add a delete entry for a file.
+   *
+   * <p>The entry's snapshot ID will be this manifest's snapshot ID. However, the original data and
+   * file sequence numbers of the file must be preserved when the file is marked as deleted.
+   *
+   * @param deletedFile a file
+   * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @param fileSequenceNumber a file sequence number (assigned when the file was added)
+   */
+  public void delete(T deletedFile, long dataSequenceNumber, Long fileSequenceNumber) {
+    currentWriter.delete(deletedFile, dataSequenceNumber, fileSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  private void tryRollingToNewFile() {
+    currentFileRows++;
+
+    if (currentWriter.length() > targetFileSizeInBytes) {
```

Comment:
Will it be expensive to call for every file? In rolling file writers, we check only every 1000 records. I understand that value may be too big but I still don't think it is a good idea to check the length for every entry.

Reply:
How about checking every 16? Just simply calculated as: `1000 / (512 / 8)`.

Reply:
Could you check the average cost of an entry in a manifest in some of your real tables? My assumption was something around 250 but let's check the average size of an entry.

Reply:
I have checked some tables. The average size has a lot to do with the schema of the table, the more columns take up more storage space. That's because of the column metrics. I noticed the biggest one (with thousand columns) needs about 5KB for one entry. The small tables (about 10 columns) need about several hundred bytes. So 250 should be an acceptable number.